Spatialite checkout with user-selected features

ymoisan opened the pull request:
I've implemented a first draft of filtering features for spatialite checkout [1].  Upon checkout, a message pops telling the user if (s)he selected features or not in each layer of the group and then the checkout proceeds extracting only user-selected features or the whole dataset in the case no feature were selected. 

This Pull Request should be viewed as a Request for Comment.  I haven't made any test yet.

Cheers,

Yves

[1] I've restricted the modifications to spatialite for now because that's the most pressing use case for us and I'm not sure how relevant it is for PG checkout

P.S. I don't know if it's my using a new editor (Atom) but you'll notice a lot of diffs pertain to spaces.  Sorry about that.

vmora commented on versioning.py:
```diff
@@ -556,10 +556,20 @@ def checkout(self):
             layer = QgsMapLayerRegistry.instance().mapLayer( layer_id )
+            layer_selected_features = layer.selectedFeatures()
+            if (layer_selected_features):
+                QMessageBox.warning(None,"Warning","You have selected "+str(len(layer_selected_features))+" features in "+layer.name()+"; checking out subset of features")
```
use self.iface.messageBar().pushMessage(...)

vmora commented:
Looks good to me, maybe making the checkout argument optional would be useful (especially since it removes the need to edit all tests).

ymoisan commented:
OK for making the argument optional (will do); for the messageBar, I know its nice and shows right up top of the main pane but I think it's important to stop with a QMessageBox so users know they are checking out a subset or not.  I fear that messages in the bar, even though the user may be forced to close them manually so he definitely has to see them, are not clear enough.  I'll check with my users what they like better.  Thanx for commenting.

ymoisan commented:
Just committed d6963b9.  We opted to stay with a MessageBox, but only for layers with selected features.  The idea is that we want to stop the user so (s)he knows a subset of data will be checked out.  For layers without any selected features, no message box pops, thereby respecting the normal plugin workflow as currently implemented, that is check out all features.  

I've made the feature_list parameter optional in versioning_base.checkout.  Is that fine with you ?

vmora commented on versioning.py:
```diff
@@ -556,10 +556,21 @@ def checkout(self):
         for layer_id in self.current_layers:
             layer = QgsMapLayerRegistry.instance().mapLayer( layer_id )
+            layer_selected_features = layer.selectedFeatures()
+            if (layer_selected_features):
```
not sure the parentheses are necessary/usefull

vmora commented on versioning.py:
```diff
@@ -712,4 +723,3 @@ def commit(self):
         else:
             QMessageBox.information(self.iface.mainWindow(), "Info",
                     "There was no modification to commit")
```
I think this should be replaced by a msg "infobar" (but not relevant for the current comit)

vmora commented on versioning_base.py:
```diff
@@ -213,14 +213,24 @@ def checkout(pg_conn_info, pg_table_names, sqlite_filename):
+                    sqlite_filename,
+                    'PG:"'+pg_conn_info+'"', schema+'.'+table,
+                    '-nln', table]
+            print ' '.join(cmd)
```
Since the feature list only adds the -where option, I'd rather have a something like:
```python
cmd = ['ogr2ogr',
                    '-preserve_fid',
                    '-f', 'SQLite',
                    '-dsco', 'SPATIALITE=yes',
                    sqlite_filename,
                    'PG:"'+pg_conn_info+'"', schema+'.'+table,
                    '-nln', table] \
                    + ['-where', '"'+pkey+' in ('+",".join([str(feature_list[i][pkey]) for i in range(0, len(feature_list))])+')"'] \
                        if feature_list else []
```

vmora commented on versioning_base.py:
```diff
@@ -235,14 +245,25 @@ def checkout(pg_conn_info, pg_table_names, sqlite_filename):
+                        sqlite_filename,
+                        'PG:"'+pg_conn_info+'"', schema+'.'+table,
+                        '-nln', table]
+
```
Same comment as above.

vmora commented:
It is personal, but I'm not found of confirmation messages. 

Have you considered modifying the checkout button icon when features are selected (maybe a yellow V sign instead of green would do), with a mouse-over help message ?

If you still prefer the msg box to this option it's not a blocker.

Please note that I'm on vacation until Nov. 2cd, so the merge should wait until then. Hope this is no blocker for you.

ymoisan commented:
Thank you for your comments Vincent.  I agree confirmation messages can be a hindrance, especially if there are many layers with subsets in the group. Modifying the icon if there are selected features or not applies to the layer group as a whole and as such just a small subset of features in one in N layers would suffice to make the checkout a "partial" one.  I

 kind of feel we need to expressly tell the user (s)he selected X features on layer Y, maybe even highlighting the features of current layer with another color while the message is displayed so the user has the time to see what is selected.  If you don't mind I'll let users tell us what they think.  

I'll make the code fixes you suggest later today.  

About your comment above that a suggestion of yours is "not relevant for the current commit", I don't mind making the plugin as a whole better, especially for low hanging fruits like what you are suggesting.  On that aspect, I hope to be able to also fix some message typos too (https://github.com/GISeHealth/qgis-versioning/issues/10).

Lastly, no problem for the PR being performed after Nov 4.  We can build the plugin for our tests.

Have a good vacation.

ymoisan commented:
I have committed b27bc50.  Summary : 
- modified ogr2ogr commands in a way similar to what you suggested
- left MenuBar code commented (in case someone prefers to uncomment those and comment out the QMessageBox ones); you can get rid of the commented lines if you wish
- did some typo editing
- modified tooltip over checkout and checkout_pg; I thought the little feather and elephant on the SVG icon files weren't all that visible and that a clearer tooltip text was needed (my 2 cents); not a blocker if you disagree

ymoisan commented:
Hi Vincent,

Looking for a way to shrink the filename displayed as info.setText in the plugin menu (GISeHealth#1), I read more comments/messages and made more typos corrections : 353f984.  Please include in current PR.

About GISeHealth#1, I was wondering if you would object to shrinking the text info shown to the user for spatialite files ?  Personally, I don't think it is important to have the full path name if what this implies is hiding plugin functionnality (icons).  I would : 

```python
#self.info.setText( uri.database() +' rev='+str(rev))
self.info.setText( uri.database().split(os.sep)[-1] +' rev='+str(rev))
```
That way the user would only see the filename.  

I don't know how much you care about showing all info.  Did you have user feedback on this ?

vmora commented:
Hi Yves,

Just pulled, make/ctest from this PR and `38% tests passed, 8 tests failed out of 13`

Am I missing something ?

ymoisan commented:
The problem was that my implementation of a default value for the optional fourth parameter in versioning_base.checkout required that the two lists (tables and features) be of equal length.  We did not have a problem in our tests within QGIS because versioning_base.checkout is always called through versioning.checkout within the context of QGIS so both lists supplied to versioning_base.checkout were always of equal length, each table having a corresponding list of selected features or an empty list (no selected feature).  

In the tests, versioning_base.checkout is called directly without supplying the fourth argument (understandably) and that revealed the issue with the default parameter value : a single empty list (length = 0).  Fixed in dfe7e8e.

vmora commented:
@ymoisan do you mind rebasing/squashing your commits and force push them on this branch ?

Ready to merge once done.

ymoisan commented:
955b99a

Thanx !
